examples/dtnerouting.rs

```diff
@@ -0,0 +1,293 @@
+use anyhow::{bail, Result};
+use clap::{crate_authors, crate_version, App, Arg};
+use dtn7::routing::erouting::ws_client::{new, Command};
+use dtn7::routing::erouting::{Packet, Sender, SenderForBundleResponse};
+use dtn7::DtnPeer;
+use futures::channel::mpsc::unbounded;
+use futures_util::{future, pin_mut, StreamExt};
+use log::{debug, info};
+use std::collections::{HashMap, HashSet};
+
+fn epi_add(history: &mut HashMap<String, HashSet<String>>, bundle_id: String, node_name: String) {
+    let entries = history.entry(bundle_id).or_insert_with(HashSet::new);
+    entries.insert(node_name);
+}
+
+fn epi_contains(
+    history: &mut HashMap<String, HashSet<String>>,
+    bundle_id: &str,
+    node_name: &str,
+) -> bool {
+    if let Some(entries) = history.get(bundle_id) {
+        debug!(
+            "Contains: {} {} {}",
+            bundle_id,
+            node_name,
+            entries.contains(node_name)
+        );
+        return entries.contains(node_name);
+    }
+    false
+}
+
+fn epi_sending_failed(
+    history: &mut HashMap<String, HashSet<String>>,
+    bundle_id: &str,
+    node_name: &str,
+) {
+    if let Some(entries) = history.get_mut(bundle_id) {
+        entries.remove(node_name);
+        debug!(
+            "removed {:?} from sent list for bundle {}",
+            node_name, bundle_id
+        );
+    }
+}
+
+fn epi_incoming_bundle(
+    history: &mut HashMap<String, HashSet<String>>,
+    bundle_id: &str,
+    node_name: &str,
+) {
+    if !node_name.is_empty() && !epi_contains(history, bundle_id, node_name) {
+        epi_add(history, bundle_id.to_string(), node_name.to_string());
+    }
+}
+
+fn epi_sending_timeout(history: &mut HashMap<String, HashSet<String>>, bundle_id: &str) {
+    if let Some(entries) = history.get_mut(bundle_id) {
+        let before = entries.len();
+        entries.clear();
+        debug!(
+            "removed {} entries from sent list for bundle {}",
+            before, bundle_id
+        );
+    }
+}
+
+#[tokio::main]
+async fn main() -> Result<()> {
```

Comment:
very high complexity in main() it should be split up into more smaller functions, especially the big loops and match blocks